index.js

```diff
@@ -129,6 +131,11 @@ function processRequestResponse (options, error, response, body) {
     /^cloudflare/i.test('' + response.caseless.get('server')) &&
     /text\/html/i.test('' + response.caseless.get('content-type'));
 
+  // If body isn't a buffer, this is a custom response body.
+  if (!Buffer.isBuffer(body)) {
```

Comment:
Should we maybe move this to the `processResponseBody` ? So the code responsible for handling the response is in one place?

Reply:
🤔 and will it be always Buffer? AFAIK it can be string as well

Reply:
> Should we maybe move this to the processResponseBody ? So the code responsible for handling the response is in one place?

The idea there was that we skip should processing of the response body altogether when it's clear that it's not a response type that we should handle. In this case, when body is not a buffer.

> will it be always Buffer? AFAIK it can be string as well

It should always be a buffer unless some other options was passed to change the response type because we're explicity setting the encoding to `null`. I'm pretty sure that `json: true` was the only exception. It might be possible for it to be a string/number but only with the `json: true` option. 

From the docs:
> encoding - encoding to be used on setEncoding of response data. If null, the body is returned as a Buffer. Anything else (including the default value of undefined) will be passed as the encoding parameter to toString() (meaning this is effectively utf8 by default). (Note: if you expect binary data, you should set encoding: null.)

Reply:
I can double check the `request` code base if you think it's necessary.

Reply:
@codemanki, I'm double checking this, one sec.

Reply:
Here is the relevant code from `request`. This function is always called because we always set a callback and is responsible for emitting the `complete` event. It is called by `Request.prototype.onRequestResponse` which sets the `response`'s encoding. Essentially `response.setEncoding(request.encoding)`. The response encoding is always `null` because we always explicitly set the `request.encoding` to `null` by passing the encoding option as `null`. This matters in the code linked below because `response` is a `readableStream` which emits the `data` event. The data emitted is ***always*** a buffer when the stream's encoding is `null`. As a matter of fact when `json: true` is based, `JSON.parse` is passed a buffer because of that. So, yes, it is certain that body will always be a buffer and that `json: true` was the only exception. Let me know if you neither any further explanation.

https://github.com/request/request/blob/7195b5064aac02532d4546c9ebcd1ec2508e6223/request.js#L1098-L1156

Reply:
@codemanki, we could even throw an error if body isn't a buffer and `options.json` isn't set because that should never happen.

Reply:
Wow! Thank you for this thorough investigation and explanation.

Reply:
Anytime :smile: